compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion/MetadataEncoding.hs

```diff
@@ -153,6 +158,53 @@ decodeOverlapMode = \case
         ]
     _ -> Nothing
 
+--------------------------
+-- INSTANCE PROPAGATION --
+--------------------------
+moduleImportsName :: LF.ExprValName
+moduleImportsName = LF.ExprValName "$$imports"
+
+encodeModuleImports :: S.Set (LF.Qualified ()) -> LF.Type
+encodeModuleImports = encodeTypeList encodeModuleImport . S.toList
+
+encodeModuleImport :: LF.Qualified () -> LF.Type
+encodeModuleImport q =
+    encodeTypeList id
+        [ encodePackageRef (LF.qualPackage q)
+        , encodeModuleName (LF.qualModule q)
+        ]
+
+encodePackageRef :: LF.PackageRef -> LF.Type
+encodePackageRef = \case
+  LF.PRSelf -> LF.TUnit
+  LF.PRImport (LF.PackageId packageId) -> TEncodedStr packageId
+
+encodeModuleName :: LF.ModuleName -> LF.Type
+encodeModuleName (LF.ModuleName components) =
+    encodeTypeList TEncodedStr components
+
+decodeModuleImports :: LF.Type -> Maybe (S.Set (LF.Qualified ()))
+decodeModuleImports = fmap S.fromList . decodeTypeList decodeModuleImport
+
+decodeModuleImport :: LF.Type -> Maybe (LF.Qualified ())
+decodeModuleImport x = decodeTypeList Just x  >>= \case
+    [packageRef, moduleName] ->
+        LF.Qualified
+            <$> decodePackageRef packageRef
+            <*> decodeModuleName moduleName
+            <*> pure ()
+    _ -> Nothing
+
+decodePackageRef :: LF.Type -> Maybe LF.PackageRef
+decodePackageRef = \case
+    LF.TUnit -> pure LF.PRSelf
+    TEncodedStr packageId -> pure (LF.PRImport (LF.PackageId packageId))
+    _ -> Nothing
+
+decodeModuleName :: LF.Type -> Maybe LF.ModuleName
+decodeModuleName = fmap LF.ModuleName . decodeTypeList \case
+    TEncodedStr component -> pure component
+    _ -> Nothing
```

Comment:
The `decodeList \case ...` is a bit hard to parse ... I think it would be nice to factor the `\case ...` part out into a separate function, `decodeText :: LF.Type -> Maybe T.Text`